File: src/test/java/org/meetkt/basket/domain/model/ItemsUnitTest.java
package org.meetkt.basket.domain.model;

import org.junit.jupiter.api.Test;
import org.meetkt.catalogue.domain.model.Product;
import org.meetkt.catalogue.domain.model.ProductId;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.meetkt.catalogue.fixture.ProductFixture.aProduct;

class ItemsUnitTest {

    @Test
    void shouldReturnAnEmptyItemGivenProductDoesNotExistInItems() {
        Product product = aProduct().withProductId("001").build();
        Items items = new Items(List.of(new Item(product)));

        Optional<Item> item = items.findFirst(new ProductId("non-existing"));

        assertThat(item).isEmpty();
    }

    @Test
    void shouldReturnAnItem() {
        Product product = aProduct().withProductId("001").build();
        Items items = new Items(List.of(new Item(product)));

        Optional<Item> item = items.findFirst(new ProductId("001"));

        assertThat(item.get()).isEqualTo(new Item(product));
    }

    @Test
    void shouldReturnTotalPrice() {
        Product product = aProduct().withProductId("001").withPriceInt(100).build();
        Product otherProduct = aProduct().withProductId("001").withPriceDouble(10.87).build();
        Items items = new Items(List.of(new Item(product), new Item(otherProduct)));

        Price totalPrice = items.totalPrice();
        assertThat(totalPrice).isEqualTo(new Price(110.87));
    }
}

File: src/test/kotlin/in/meetkt/basket/domain/model/ItemsUnitTest.kt
package `in`.meetkt.basket.domain.model

import `in`.meetkt.catalogue.domain.model.ProductId
import `in`.meetkt.catalogue.fixture.ProductFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ItemsUnitTest {

    @Test
    fun shouldReturnAnEmptyItemGivenProductDoesNotExistInItems() {
        val product = ProductFixture.aProduct().withProductId("001").build()
        val items = Items(listOf(Item(product)))
        val item = items.findFirst(ProductId("non-existing"))

        assertThat(item).isNull()
    }

    @Test
    fun shouldReturnAnItem() {
        val product = ProductFixture.aProduct().withProductId("001").build()
        val items = Items(listOf(Item(product)))
        val item = items.findFirst(ProductId("001"))

        assertThat(item).isEqualTo(Item(product))
    }

    @Test
    fun shouldReturnTotalPrice() {
        val product =
            ProductFixture.aProduct().withProductId("001").withPriceInt(100).build()
        val otherProduct =
            ProductFixture.aProduct().withProductId("001").withPriceDouble(10.87).build()
        val items = Items(
            listOf(
                Item(product),
                Item(otherProduct)
            )
        )
        val totalPrice = items.totalPrice()
        assertThat(totalPrice).isEqualTo(Price(110.87))
    }
}